### ISSUE TITLE
importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated

### ISSUE THREAD

**Issue by kzidane** [cs50/check50#111]
In Python 3.7, `cs50/2018/x/finance`:

```
 check50 ran into an error while running checks!
    sending GET request to /login
    Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/unittest/case.py", line 615, in run
    testMethod()
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/check50.py", line 416, in wrapper
    func(self)
      File "/root/.local/share/check50/cs50/checks/cs50/2018/x/finance/check50/__init__.py", line 115, in login_page
    self.validate_form("/login", ["username", "password"])
      File "/root/.local/share/check50/cs50/checks/cs50/2018/x/finance/check50/__init__.py", line 10, in wrapper
    return f(self, *args, **kwargs)
      File "/root/.local/share/check50/cs50/checks/cs50/2018/x/finance/check50/__init__.py", line 35, in validate_form
    content = self.app.get(route).content()
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/check50.py", line 520, in content
    BeautifulSoup(self.response.data, "html.parser"),
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/__init__.py", line 282, in __init__
    self._feed()
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/__init__.py", line 343, in _feed
    self.builder.feed(self.markup)
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/builder/_htmlparser.py", line 247, in feed
    parser.feed(markup)
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/html/parser.py", line 111, in feed
    self.goahead(0)
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/html/parser.py", line 179, in goahead
    k = self.parse_html_declaration(i)
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/html/parser.py", line 270, in parse_html_declaration
    self.handle_decl(rawdata[i+2:gtpos])
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/builder/_htmlparser.py", line 192, in handle_decl
    self.soup.endData(Doctype)
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/__init__.py", line 420, in endData
    self.object_was_parsed(o)
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/__init__.py", line 425, in object_was_parsed
    previous_element = most_recent_element or self._most_recent_element
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/element.py", line 1117, in __getattr__
    return self.find(tag)
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/element.py", line 1354, in find
    l = self.find_all(name, attrs, recursive, text, 1, **kwargs)
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/element.py", line 1375, in find_all
    return self._find_all(name, attrs, text, limit, generator, **kwargs)
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/element.py", line 582, in _find_all
    strainer = SoupStrainer(name, attrs, text, **kwargs)
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/element.py", line 1686, in __init__
    self.text = self._normalize_search_value(text)
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/site-packages/bs4/element.py", line 1691, in _normalize_search_value
    if (isinstance(value, str) or isinstance(value, collections.Callable) or hasattr(value, 'match')
      File "/opt/pyenv/versions/3.7.0/lib/python3.7/collections/__init__.py", line 52, in __getattr__
    DeprecationWarning, stacklevel=2)
    Contact sysadmins@cs50.harvard.edu with the URL of this check!
```

@crossroads1112 @Jelleas any workarounds or should we downgrade back to the latest 3.6?

See also https://github.com/pypa/setuptools/issues/1401.

**Comment by Jelleas** [comment]
Can't replicate this. Seems to be a bug in bs4 that was recently patched (https://bugs.launchpad.net/beautifulsoup/+bug/1778909). Try updating bs4?

**Comment by kzidane** [comment]
@Jelleas what Python version are you using? This happens on 3.7. I think I have the latest version of `bs4`:

```
$ pip show bs4
Name: bs4
Version: 0.0.1
Summary: Screen-scraping library
Home-page: https://pypi.python.org/pypi/beautifulsoup4
Author: Leonard Richardson
Author-email: leonardr@segfault.org
License: MIT
Location: /home/kzidane/.virtualenvs/check50/lib/python3.6/site-packages
Requires: beautifulsoup4
Required-by: check50
$ pip install --upgrade bs4
Requirement already up-to-date: bs4 in /home/kzidane/.virtualenvs/check50/lib/python3.6/site-packages (0.0.1)
Requirement not upgraded as not directly required: beautifulsoup4 in /home/kzidane/.virtualenvs/check50/lib/python3.6/site-packages (from bs4) (4.6.0)
```

**Comment by kzidane** [comment]
Ah, that's my fault. I was testing on different python versions and forgot to switch to 3.7 before I ran the commands above to show you the output.

```
$ pip install --upgrade bs4
Requirement already up-to-date: bs4 in /home/kzidane/.virtualenvs/check50/lib/python3.7/site-packages (0.0.1)
Requirement already satisfied, skipping upgrade: beautifulsoup4 in /home/kzidane/.virtualenvs/check50/lib/python3.7/site-packages (from bs4) (4.6.1)
```

**Comment by dmalan** [comment]
@kzidane just to be clear, 4.6.1 (released 2 days ago) doesn't, in fact, solve, per the bottommost comment at https://bugs.launchpad.net/beautifulsoup/+bug/1778909? Cf. https://pypi.org/project/beautifulsoup4/#history

**Comment by dmalan** [comment]
@curiouskiwi

**Comment by crossroads1112** [comment]
Solved in the short term by #112 which silences the warning.
